packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -270,19 +273,38 @@ export function esbuildCjsExternalPlugin(externals: string[]): Plugin {
         `^${text.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')}$`
       const filter = new RegExp(externals.map(escape).join('|'))
 
-      build.onResolve({ filter: /.*/, namespace: 'external' }, (args) => ({
-        path: args.path,
-        external: true
-      }))
+      build.onResolve({ filter: new RegExp(`^${nonFacadePrefix}`) }, (args) => {
+        return {
+          path: args.path.slice(nonFacadePrefix.length),
+          external: true
+        }
+      })
+
+      build.onResolve({ filter }, (args) => {
+        if (args.kind === 'require-call') {
+          return {
+            path: args.path,
+            namespace: cjsExternalFacadeNamespace
+          }
+        }
 
-      build.onResolve({ filter }, (args) => ({
-        path: args.path,
-        namespace: 'external'
-      }))
+        return {
+          path: args.path,
+          external: true
+        }
+      })
 
-      build.onLoad({ filter: /.*/, namespace: 'external' }, (args) => ({
-        contents: `export * from ${JSON.stringify(args.path)}`
-      }))
+      build.onLoad(
+        { filter: /.*/, namespace: cjsExternalFacadeNamespace },
+        (args) => ({
+          contents:
+            `import * as m from ${JSON.stringify(
+              nonFacadePrefix + args.path
+            )};` +
+            `export default m.default;` +
+            `export * from ${JSON.stringify(nonFacadePrefix + args.path)};`
```

Comment:
This part is not correct and is not possible to do it correctly.
In Node environment, it is possible by using `const require = createRequire(import.meta.url)`. I guess this will cover @jiby-aurum's case.
For browsers, I'm not sure how we should handle this one.

Reply:
I haven't thought of this, but yes. In ssr mode we should avoid rewriting require to import for node built-in modules, but rather make `require` available via `createRequire`.

Reply:
I'm not sure what's not correct here, except that we always return a default export regardless if it has or not? which I think is a good tradeoff for now to get most things working.

Reply:
> I'm not sure what's not correct here

`const foo = require('foo')` will be converted to `import * as foo from 'foo'`. But this won't work for every package. That said, I believe it's ok since externalizing required dependencies aren't recommended.